File: internal/laporanpertandingan/service.go
package laporanpertandingan

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

type Service struct {
	repo *Repository
}

func NewService(repo *Repository) *Service {
	return &Service{repo: repo}
}

// GetAll ambil semua laporan dengan filter opsional
func (s *Service) GetAll(filter FilterLaporan) ([]LaporanDetail, error) {
	rows, err := s.repo.GetAll(filter)
	if err != nil {
		return nil, err
	}
	// Populate atlet per row
	for i := range rows {
		atletA, _ := s.repo.GetAtletBySisi(rows[i].ID, "A")
		atletB, _ := s.repo.GetAtletBySisi(rows[i].ID, "B")
		if atletA == nil {
			atletA = []AtletSisiItem{}
		}
		if atletB == nil {
			atletB = []AtletSisiItem{}
		}
		rows[i].AtletA = atletA
		rows[i].AtletB = atletB
	}
	return rows, nil
}

// GetByID ambil satu laporan lengkap dengan atlet
func (s *Service) GetByID(id uint) (*LaporanDetail, error) {
	laporan, err := s.repo.GetByID(id)
	if err != nil {
		return nil, errors.New("laporan pertandingan tidak ditemukan")
	}

	atletA, _ := s.repo.GetAtletBySisi(id, "A")
	atletB, _ := s.repo.GetAtletBySisi(id, "B")
	if atletA == nil {
		atletA = []AtletSisiItem{}
	}
	if atletB == nil {
		atletB = []AtletSisiItem{}
	}
	laporan.AtletA = atletA
	laporan.AtletB = atletB

	return laporan, nil
}

// Create buat laporan baru + insert atlet per sisi
func (s *Service) Create(req *CreateLaporanRequest, userID uint) (*LaporanDetail, error) {
	// Normalisasi ke uppercase agar toleran terhadap input frontend
	req.Babak = strings.ToUpper(req.Babak)
	req.Pemenang = strings.ToUpper(req.Pemenang)

	if err := validateBabak(req.Babak); err != nil {
		return nil, err
	}
	if err := validatePemenang(req.Pemenang); err != nil {
		return nil, err
	}

	var createdBy *uint
	if userID > 0 {
		createdBy = &userID
	}

	laporan := &LaporanPertandingan{
		TanggalPertandingan: TanggalDate{mustParseDate(req.TanggalPertandingan)},
		WaktuPertandingan:   req.WaktuPertandingan,
		Venue:               req.Venue,
		CaborID:             req.CaborID,
		NomorID:             req.NomorID,
		Babak:               req.Babak,
		KontingenAID:        req.KontingenAID,
		KontingenBID:        req.KontingenBID,
		HasilPertandingan:   req.HasilPertandingan,
		Pemenang:            req.Pemenang,
		JuaraKe:             req.JuaraKe,
		Wasit:               req.Wasit,
		CatatanKhusus:       req.CatatanKhusus,
		CreatedBy:           createdBy,
	}

	if err := s.repo.Create(laporan); err != nil {
		return nil, fmt.Errorf("gagal membuat laporan: %w", err)
	}

	// Insert atlet sisi A dan B
	if len(req.AtletA) > 0 {
		if err := s.repo.ReplaceAtletSisi(laporan.ID, "A", req.AtletA); err != nil {
			return nil, fmt.Errorf("gagal menyimpan atlet sisi A: %w", err)
		}
	}
	if len(req.AtletB) > 0 {
		if err := s.repo.ReplaceAtletSisi(laporan.ID, "B", req.AtletB); err != nil {
			return nil, fmt.Errorf("gagal menyimpan atlet sisi B: %w", err)
		}
	}

	return s.GetByID(laporan.ID)
}

// Update ubah data laporan — partial, hanya field yang dikirim
func (s *Service) Update(id uint, req *UpdateLaporanRequest) (*LaporanDetail, error) {
	// Pastikan laporan ada
	if _, err := s.repo.GetByID(id); err != nil {
		return nil, errors.New("laporan pertandingan tidak ditemukan")
	}

	updates := make(map[string]interface{})
	if req.TanggalPertandingan != "" {
		updates["tanggal_pertandingan"] = mustParseDate(req.TanggalPertandingan)
	}
	if req.WaktuPertandingan != "" {
		updates["waktu_pertandingan"] = req.WaktuPertandingan
	}
	if req.Venue != "" {
		updates["venue"] = req.Venue
	}
	if req.CaborID > 0 {
		updates["cabor_id"] = req.CaborID
	}
	if req.NomorID > 0 {
		updates["nomor_id"] = req.NomorID
	}
	if req.Babak != "" {
		req.Babak = strings.ToUpper(req.Babak)
		if err := validateBabak(req.Babak); err != nil {
			return nil, err
		}
		updates["babak"] = req.Babak
	}
	if req.KontingenAID > 0 {
		updates["kontingen_a_id"] = req.KontingenAID
	}
	if req.KontingenBID != nil {
		updates["kontingen_b_id"] = req.KontingenBID
	}
	if req.HasilPertandingan != "" {
		updates["hasil_pertandingan"] = req.HasilPertandingan
	}
	if req.Pemenang != "" {
		req.Pemenang = strings.ToUpper(req.Pemenang)
		if err := validatePemenang(req.Pemenang); err != nil {
			return nil, err
		}
		updates["pemenang"] = req.Pemenang
	}
	if req.JuaraKe != nil {
		updates["juara_ke"] = req.JuaraKe
	}
	if req.Wasit != "" {
		updates["wasit"] = req.Wasit
	}
	if req.CatatanKhusus != nil {
		updates["catatan_khusus"] = req.CatatanKhusus
	}

	if len(updates) > 0 {
		if err := s.repo.Update(id, updates); err != nil {
			return nil, fmt.Errorf("gagal mengupdate laporan: %w", err)
		}
	}

	// Update atlet jika dikirim (replace)
	if req.AtletA != nil {
		if err := s.repo.ReplaceAtletSisi(id, "A", req.AtletA); err != nil {
			return nil, fmt.Errorf("gagal update atlet sisi A: %w", err)
		}
	}
	if req.AtletB != nil {
		if err := s.repo.ReplaceAtletSisi(id, "B", req.AtletB); err != nil {
			return nil, fmt.Errorf("gagal update atlet sisi B: %w", err)
		}
	}

	return s.GetByID(id)
}

// Delete hapus laporan (atlet ikut terhapus via ON DELETE CASCADE)
func (s *Service) Delete(id uint) error {
	if _, err := s.repo.GetByID(id); err != nil {
		return errors.New("laporan pertandingan tidak ditemukan")
	}
	return s.repo.Delete(id)
}

// UpdateFile update path foto_bukti atau video_bukti
func (s *Service) UpdateFile(id uint, column, path string) error {
	if _, err := s.repo.GetByID(id); err != nil {
		return errors.New("laporan pertandingan tidak ditemukan")
	}
	return s.repo.UpdateFile(id, column, path)
}

// mustParseDate parse "YYYY-MM-DD" ke time.Time. Return zero jika gagal.
func mustParseDate(s string) time.Time {
	for _, layout := range []string{"2006-01-02", time.RFC3339} {
		if t, err := time.Parse(layout, s); err == nil {
			return t
		}
	}
	return time.Time{}
}

// ===== VALIDASI ENUM =====

var validBabak = map[string]bool{
	"PENYISIHAN": true, "8_BESAR": true, "PEREMPAT_FINAL": true,
	"SEMIFINAL": true, "FINAL": true, "PEREBUTAN_TEMPAT_3": true, "LAINNYA": true,
}

var validPemenang = map[string]bool{
	"TIM_A": true, "TIM_B": true, "DRAW": true,
}

func validateBabak(b string) error {
	if !validBabak[b] {
		return errors.New("babak tidak valid (PENYISIHAN/8_BESAR/PEREMPAT_FINAL/SEMIFINAL/FINAL/PEREBUTAN_TEMPAT_3/LAINNYA)")
	}
	return nil
}

func validatePemenang(p string) error {
	if !validPemenang[p] {
		return errors.New("pemenang tidak valid (TIM_A/TIM_B/DRAW)")
	}
	return nil
}

// ===== DROPDOWN =====

// GetKontingenDropdown ambil semua kontingen untuk dropdown Tim A/B
func (s *Service) GetKontingenDropdown() ([]KontingenDropdownItem, error) {
	return s.repo.GetKontingenDropdown()
}

// GetCaborDropdown ambil cabor aktif untuk dropdown
func (s *Service) GetCaborDropdown() ([]CaborDropdownItem, error) {
	return s.repo.GetCaborDropdown()
}

// GetNomorDropdown ambil nomor aktif, difilter by cabor_id jika ada
func (s *Service) GetNomorDropdown(caborID uint) ([]NomorDropdownItem, error) {
	return s.repo.GetNomorDropdown(caborID)
}

// GetAtletTerdaftarDropdown ambil atlet dari trx_pendaftaran_atlet
// Filter: kontingen_id, cabor_id, nomor_id — semua opsional
func (s *Service) GetAtletTerdaftarDropdown(kontingenID, caborID, nomorID uint) ([]AtletTerdaftarDropdownItem, error) {
	return s.repo.GetAtletTerdaftarDropdown(kontingenID, caborID, nomorID)
}

File: LaporanPertandinganFrontend/service.ts
import api from "../../lib/api";

const BASE_URL = "http://localhost:8000";
const getToken = () => localStorage.getItem("token") ?? "";

// ─── Enums ────────────────────────────────────────────────

export type Babak =
  | "PENYISIHAN"
  | "8_BESAR"
  | "PEREMPAT_FINAL"
  | "SEMIFINAL"
  | "FINAL"
  | "PEREBUTAN_TEMPAT_3"
  | "LAINNYA";

export type Pemenang = "TIM_A" | "TIM_B" | "DRAW";

export const BABAK_OPTIONS: { value: Babak; label: string }[] = [
  { value: "PENYISIHAN",         label: "Penyisihan" },
  { value: "8_BESAR",            label: "8 Besar" },
  { value: "PEREMPAT_FINAL",     label: "Perempat Final" },
  { value: "SEMIFINAL",          label: "Semifinal" },
  { value: "FINAL",              label: "Final" },
  { value: "PEREBUTAN_TEMPAT_3", label: "Perebutan Tempat 3" },
  { value: "LAINNYA",            label: "Lainnya" },
];

export const PEMENANG_OPTIONS: { value: Pemenang; label: string }[] = [
  { value: "TIM_A", label: "Tim A" },
  { value: "TIM_B", label: "Tim B" },
  { value: "DRAW",  label: "Seri" },
];

// ─── Types ────────────────────────────────────────────────

export interface AtletSisiItem {
  id: number;
  atlet_id: number;
  nama_lengkap: string;
  urutan: number;
}

export interface LaporanDetail {
  id: number;
  tanggal_pertandingan: string;   // YYYY-MM-DD
  waktu_pertandingan: string;     // HH:MM:SS
  venue: string;
  cabor_id: number;
  nomor_id: number;
  babak: Babak;
  kontingen_a_id: number;
  kontingen_b_id: number | null;
  hasil_pertandingan: string;
  pemenang: Pemenang;
  juara_ke: number | null;
  wasit: string;
  catatan_khusus: string | null;
  foto_bukti: string | null;
  video_bukti: string | null;
  created_by: number | null;
  created_at: string;
  updated_at: string;
  // Join fields
  nama_cabor: string;
  nama_nomor: string;
  nama_kontingen_a: string;
  nama_kontingen_b: string | null;
  atlet_a: AtletSisiItem[];
  atlet_b: AtletSisiItem[];
}

/** Payload POST — jangan kirim created_by, diisi otomatis JWT */
export interface CreateLaporanPayload {
  tanggal_pertandingan: string;   // YYYY-MM-DD, wajib
  waktu_pertandingan: string;     // HH:MM:SS, wajib
  venue: string;                  // wajib
  cabor_id: number;               // wajib
  nomor_id: number;               // wajib
  babak: Babak;                   // wajib
  kontingen_a_id: number;         // wajib
  kontingen_b_id?: number;        // opsional
  hasil_pertandingan: string;     // wajib
  pemenang: Pemenang;             // wajib
  juara_ke?: number;              // opsional
  wasit: string;                  // wajib
  catatan_khusus?: string;        // opsional
  atlet_a?: number[];             // ordered list atlet_id sisi A
  atlet_b?: number[];             // ordered list atlet_id sisi B
}

/** Payload PUT — partial update */
export interface UpdateLaporanPayload {
  tanggal_pertandingan?: string;
  waktu_pertandingan?: string;
  venue?: string;
  cabor_id?: number;
  nomor_id?: number;
  babak?: Babak;
  kontingen_a_id?: number;
  kontingen_b_id?: number | null;
  hasil_pertandingan?: string;
  pemenang?: Pemenang;
  juara_ke?: number | null;
  wasit?: string;
  catatan_khusus?: string | null;
  /**
   * Replace semua atlet sisi A. Tidak dikirim = tidak berubah.
   * Array kosong [] = hapus semua atlet sisi A.
   */
  atlet_a?: number[];
  /**
   * Replace semua atlet sisi B. Tidak dikirim = tidak berubah.
   * Array kosong [] = hapus semua atlet sisi B.
   */
  atlet_b?: number[];
}

export interface LaporanFilter {
  tanggal?: string;     // YYYY-MM-DD
  cabor_id?: number;
  nomor_id?: number;
  babak?: Babak;
  pemenang?: Pemenang;
}

/** Data satu penandatangan untuk export PDF */
export interface TTDData {
  jabatan: string;
  nama_tercetak: string;
  nip?: string;
  signature_b64?: string; // base64 PNG dari signature pad
}

/** Payload export PDF — batch maupun single */
export interface ExportPDFPayload {
  tanggal?: string;         // filter per hari (batch only)
  cabor_id?: number;        // filter cabor (batch only)
  nomor_id?: number;        // filter nomor (batch only)
  penandatangan?: TTDData[];
}

// ─── Service ─────────────────────────────────────────────

/**
 * GET /admin/laporan-pertandingan
 * List semua laporan, diurutkan tanggal DESC, waktu DESC.
 */
export const getAllLaporan = (
  filters?: LaporanFilter
): Promise<LaporanDetail[]> =>
  api.get("/admin/laporan-pertandingan", { params: filters }).then(r => r.data.data ?? []);

/**
 * GET /admin/laporan-pertandingan/:id
 * Detail satu laporan lengkap termasuk atlet_a dan atlet_b.
 */
export const getLaporanById = (id: number): Promise<LaporanDetail> =>
  api.get(`/admin/laporan-pertandingan/${id}`).then(r => r.data.data);

/**
 * POST /admin/laporan-pertandingan
 * Buat laporan baru. created_by otomatis dari JWT — tidak perlu dikirim.
 */
export const createLaporan = (
  payload: CreateLaporanPayload
): Promise<LaporanDetail> =>
  api.post("/admin/laporan-pertandingan", payload).then(r => r.data.data);

/**
 * PUT /admin/laporan-pertandingan/:id
 * Partial update. Untuk atlet: jika field tidak ada di body → tidak berubah.
 * Kirim [] → hapus semua atlet sisi tersebut.
 */
export const updateLaporan = (
  id: number,
  payload: UpdateLaporanPayload
): Promise<LaporanDetail> =>
  api.put(`/admin/laporan-pertandingan/${id}`, payload).then(r => r.data.data);

/**
 * DELETE /admin/laporan-pertandingan/:id
 * Hard delete. Data atlet (laporan_pertandingan_atlet) ikut terhapus via CASCADE.
 */
export const deleteLaporan = (
  id: number
): Promise<{ success: boolean; message: string }> =>
  api.delete(`/admin/laporan-pertandingan/${id}`).then(r => r.data);

/**
 * PUT /admin/laporan-pertandingan/:id/foto
 * Upload foto bukti. Field name di FormData harus "foto".
 */
export const uploadFotoLaporan = async (
  id: number,
  file: File
): Promise<{ success: boolean; message: string; path: string }> => {
  const fd = new FormData();
  fd.append("foto", file);
  const r = await fetch(`${BASE_URL}/admin/laporan-pertandingan/${id}/foto`, {
    method: "PUT",
    headers: { Authorization: `Bearer ${getToken()}` },
    body: fd,
  });
  const data = await r.json().catch(() => ({}));
  if (!r.ok || data.success === false)
    throw new Error(data.message || `Error ${r.status}`);
  return data;
};

/**
 * PUT /admin/laporan-pertandingan/:id/video
 * Upload video bukti. Field name di FormData harus "video".
 */
export const uploadVideoLaporan = async (
  id: number,
  file: File
): Promise<{ success: boolean; message: string; path: string }> => {
  const fd = new FormData();
  fd.append("video", file);
  const r = await fetch(`${BASE_URL}/admin/laporan-pertandingan/${id}/video`, {
    method: "PUT",
    headers: { Authorization: `Bearer ${getToken()}` },
    body: fd,
  });
  const data = await r.json().catch(() => ({}));
  if (!r.ok || data.success === false)
    throw new Error(data.message || `Error ${r.status}`);
  return data;
};

/**
 * POST /admin/laporan-pertandingan/:id/export/pdf
 * Export PDF satu pertandingan. Body opsional (tanpa body = tanpa tanda tangan).
 * Response: file PDF binary (blob).
 */
export const exportSatuPDF = async (
  id: number,
  payload: ExportPDFPayload = {}
): Promise<void> => {
  const res = await fetch(`${BASE_URL}/admin/laporan-pertandingan/${id}/export/pdf`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.message || `Error ${res.status}`);
  }
  const blob = await res.blob();
  const a = document.createElement("a");
  a.href = URL.createObjectURL(blob);
  a.download = `laporan_${id}.pdf`;
  a.click();
  URL.revokeObjectURL(a.href);
};

/**
 * POST /admin/laporan-pertandingan/export/pdf
 * Export PDF batch (semua / per hari / per cabor).
 * Pakai POST karena body bisa berisi data tanda tangan base64.
 * PENTING: route statis /export/pdf terdaftar sebelum /:id di backend.
 */
export const exportBatchPDF = async (
  payload: ExportPDFPayload = {}
): Promise<void> => {
  const res = await fetch(`${BASE_URL}/admin/laporan-pertandingan/export/pdf`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.message || `Error ${res.status}`);
  }
  const blob = await res.blob();
  const a = document.createElement("a");
  a.href = URL.createObjectURL(blob);
  const suffix = payload.tanggal ?? "semua";
  a.download = `laporan_pertandingan_${suffix}.pdf`;
  a.click();
  URL.revokeObjectURL(a.href);
};

// ─── Service object ───────────────────────────────────────

export const laporanPertandinganService = {
  getAll:       getAllLaporan,
  getById:      getLaporanById,
  create:       createLaporan,
  update:       updateLaporan,
  delete:       deleteLaporan,
  uploadFoto:   uploadFotoLaporan,
  uploadVideo:  uploadVideoLaporan,
  exportSatu:   exportSatuPDF,
  exportBatch:  exportBatchPDF,
};
